### PR TITLE
[quic]Check canonical suffix when checking checking QUIC brokenness

### PR DESCRIPTION
**Pull request by RenjieTang**
Commit Message: Check canonical suffix when checking checking QUIC brokenness
Additional Description: 
Risk Level: low. protected by default false runtime guard
Testing: unit test
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a
Runtime guard: envoy_reloadable_features_use_canonical_suffix_for_quic_broken

### CODE REVIEW

**Comment by RenjieTang** [comment]
@abeyad @fredyw for visibility.

File: test/common/http/http_server_properties_cache_impl_test.cc
Comment (fredyw) [inline test/common/http/http_server_properties_cache_impl_test.cc:427]: Can we also update `CanonicalSuffixNoMatch` test case by marking `origin1` as broken and asserting that `origin2` is not broken?

File: source/common/runtime/runtime_features.cc
Comment (fredyw) [inline source/common/runtime/runtime_features.cc:156]: nit: `_brokenness`

File: source/common/http/http_server_properties_cache_impl.cc
Comment (fredyw) [inline source/common/http/http_server_properties_cache_impl.cc:311]: nit: You can use if expression, i.e. `if (auto entry_it = protocols_.find(origin); ....) { ... }`

File: source/common/http/http_server_properties_cache_impl.cc
Comment (fredyw) [inline source/common/http/http_server_properties_cache_impl.cc:316]: nit: ditto

File: envoy/http/http_server_properties_cache.h
Comment (fredyw) [inline envoy/http/http_server_properties_cache.h:180]: typo: `@return true if it's broken`

File: test/common/http/http_server_properties_cache_impl_test.cc
Comment (RenjieTang) [reply test/common/http/http_server_properties_cache_impl_test.cc:427]: done

File: source/common/http/http_server_properties_cache_impl.cc
Comment (danzh2010) [inline source/common/http/http_server_properties_cache_impl.cc:300]: Sorry, I misread the implementation of getOrCreateHttp3StatusTracker() above. setPropertiesImpl() actually update canonical suffix mapping with the new origin. So even without your fix in isHttp3Broken(), CanonicalSuffix test (with marking origin1 broken and checking origin2 broken) may still pass. But if you mark origin2 broken and assert origin1 broken, it will fail. (However, I'm not sure if this is the desired behavior or origin 1 and origin2 brokenness should be consistant.) @RyanTheOptimist

**Review by fredyw — APPROVED** [review]

File: source/common/http/http_server_properties_cache_impl.cc
Comment (RenjieTang) [reply source/common/http/http_server_properties_cache_impl.cc:300]: I ran the test without flipping my flag and it failed.
In that case I think what happens is that we call getOrCreateHttp3StatusTracker() and fail to find a tracker. We create a default tracker with h3_status as pending. So h3 is not considered broken.

Comment (danzh2010) [reply source/common/http/http_server_properties_cache_impl.cc:300]: Can you test the case where origin2 is marked broken and check origin1?

Comment (RenjieTang) [reply source/common/http/http_server_properties_cache_impl.cc:300]: Whether the flag is flipped or not, marking origin2 broken doesn't affect origin1. origin1 returns false for h3 brokenness. 

When the flag is false, origin1 is not considered broken because we don't check for canonial suffix. When the flag is true, origin1 is not broken because we don't create a new alt_protocol and thus not update the canonical suffix.

Comment (danzh2010) [reply source/common/http/http_server_properties_cache_impl.cc:300]: > Whether the flag is flipped or not, marking origin2 broken doesn't affect origin1. origin1 returns false for h3 brokenness.
> 

I'm not sure if this is desired or not. If we add alternative protocols for x.example.com, and then check shouldAttemptHttp3() on y.example.com, it returns yes. And then if we mark y.example.com HTTP/3 broken, x.example.com is regarded not broken.

Comment (RenjieTang) [reply source/common/http/http_server_properties_cache_impl.cc:300]: I think in prod, if we are to mark `y.example.com` broken, we will actually [create](https://github.com/envoyproxy/envoy/blob/44c5b009567b322b7599f600342e6d4152e87ecd/source/common/http/conn_pool_grid.cc#L495) a status tracker and set the canonical suffix. This will prevent `z.example.com` from attempting QUIC, but it won't prevent `x.example.com` because we will find its exact record so we don't check canonical suffix. Should I change the logic to only lookup canonical suffix but not the exact record?

Comment (danzh2010) [reply source/common/http/http_server_properties_cache_impl.cc:300]: > This will prevent z.example.com from attempting QUIC

I'm not sure about this, I would have thought that canonical suffix (i.e. example.com) is tracking x.example.com, and following z.example.com will still attempt QUIC. Would love to see a test for this.

**Comment by RenjieTang** [comment]
I reworked this PR to have canonical suffixes of broken HTTP/3 origins separately tracked. PTAL

**Review by RyanTheOptimist — DISMISSED** [review]
LGTM, but please wait for Dan's approval too

File: test/common/http/http_server_properties_cache_impl_test.cc
Comment (danzh2010) [inline test/common/http/http_server_properties_cache_impl_test.cc:485]: See my earlier comment, I would think we want to treat origin1 as broken too if origin2 is already marked broken.

**Review by danzh2010 — COMMENTED** [review]
I think this is a neat approach! One more detail to discuss in comment.

File: source/common/http/http_server_properties_cache_impl.cc
Comment (RenjieTang) [reply source/common/http/http_server_properties_cache_impl.cc:325]: If we have an initial quic hint, it will be in `protocol_`, but the entry won't have a h3 tracker until either markH3Confirmed() or markH3Broken() is called.

File: source/common/http/http_server_properties_cache_impl.cc
Comment (RenjieTang) [reply source/common/http/http_server_properties_cache_impl.cc:315]: Good point. Updated the code and also test.

File: test/common/http/http_server_properties_cache_impl_test.cc
Comment (RenjieTang) [reply test/common/http/http_server_properties_cache_impl_test.cc:517]: Done updating. Note that there are 2 cases. When origin2 fails but origin1 doesn't have any h3 status, later origin1 requests will consider h3 broken.
But if origin2 fails after origin1 is confirmed, origin1 will keep its confirmed status

Comment (danzh2010) [reply test/common/http/http_server_properties_cache_impl_test.cc:517]: These expectations make sense to me! Thanks for adding tests!

File: source/common/http/http_server_properties_cache_impl.cc
Comment (danzh2010) [reply source/common/http/http_server_properties_cache_impl.cc:326]: After using a different map canonical_h3_broken_map_ to track canonical origin for H3 brokenness, if getHttp3BrokenCanonicalOrigin() doesn't return nullopt, the canonical origin must have a valid H3 status, right?  IIUC, the new map is only populated when we markH3Confirmed() or markH3Broken().

File: source/common/http/http_server_properties_cache_impl.cc
Comment (danzh2010) [reply source/common/http/http_server_properties_cache_impl.cc:315]: Please also update the comment above that we don't create status tracker in any case now.

File: source/common/http/http_server_properties_cache_impl.cc
Comment (RenjieTang) [reply source/common/http/http_server_properties_cache_impl.cc:325]: You are right. If an origin exists in the canonical_h3_broken_map_, it must have a http3 tracker. What would you suggest me to do here? Just use the tracker without checking for nullptr?

File: source/common/http/http_server_properties_cache_impl.h
Comment (danzh2010) [reply source/common/http/http_server_properties_cache_impl.h:164]: nits: Please update the comment.

**Review by danzh2010 — APPROVED** [review]
LGTM

File: source/common/http/http_server_properties_cache_impl.h
Comment (RenjieTang) [reply source/common/http/http_server_properties_cache_impl.h]: done.

**Review by danzh2010 — APPROVED** [review]

**Review by RyanTheOptimist — APPROVED** [review]

**Pull request merged**
